scipy/special/_logsumexp.py

```diff
@@ -100,53 +101,113 @@ def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
     """
     xp = array_namespace(a, b)
     a, b = xp_broadcast_promote(a, b, ensure_writeable=True, force_floating=True, xp=xp)
-    axis = tuple(range(a.ndim)) if axis is None else axis
+    a = xp_atleast_nd(a, ndim=1)
+    b = xp_atleast_nd(b, ndim=1) if b is not None else b
 
-    if b is not None:
-        a[b == 0] = -xp.inf
-
-    # Scale by real part for complex inputs, because this affects
-    # the magnitude of the exponential.
     if xp_size(a) == 0:
-        # because `xp.max` doesn't have `initial` argument...
-        shape = np.asarray(a.shape)  # NumPy is concise for scalar or tuple `axis`
+        shape = np.asarray(a.shape)  # NumPy is convenient for shape manipulation
+        axis = tuple(range(a.ndim)) if axis is None else axis
         shape[axis] = 1
-        a_max = xp.full(tuple(shape), -xp.inf, dtype=a.dtype)
+        out = xp.full(tuple(shape), -xp.inf, dtype=a.dtype)
+        sgn = xp.sign(out)
+    else:
+        with np.errstate(divide='ignore', invalid='ignore'):  # log of zero is OK
+            out, sgn = _logsumexp(a, b, axis=axis, return_sign=return_sign, xp=xp)
```

Comment:
I'd have swapped these lines so the most common behavior occurs first, and the exceptional case (`xp_size(a) == 0`) comes later. It took me a bit to figure out what the function is doing!

Reply:
I can swap these for you if you prefer.